src/components/Topology/Topology.tsx

```diff
@@ -141,206 +140,216 @@ const getRelationPosition = (data: string[]) => {
   };
 };
 
-const Topology = ({
-  annotations,
-  applications: applicationData,
-  relations: relationData,
-  width,
-  height,
-}: Props) => {
-  const ref = useRef<SVGSVGElement>(null);
-
-  const annotationData = cloneDeep(annotations) ?? {};
-
-  const { deltaX, deltaY } = computePositionDelta(annotationData);
-
-  const applications: Application[] = Object.entries(applicationData ?? {}).map(
-    ([appName, application]) => {
+// Memoize the component as the D3 rendering is expensive and causes visual
+// glitches.
+const Topology = memo(
```

Comment:
Wrapping this component in `memo()` is the only change, everything else is just indentation changes.